### DISCUSSION TITLE
Logic For Collapsing in VersionedCollapsingMergeTree Engine

### DISCUSSION THREAD

**Discussion by Sourish-Mukherjee** [ClickHouse/ClickHouse#66805] [Q&A]
I am having a confusion, in how clickhouse is collapsing rows with the above said engine.

Lets say i create table

```
CREATE TABLE table1
(
    ID UInt64,
    Name String,
    Col1 Bool,
    Sign Int8,
    Version UInt8
)
ENGINE = VersionedCollapsingMergeTree(Sign, Version)
PRIMARY KEY (ID)
ORDER BY (ID, Name);
```

Now i am inserting into table with **one-by-one** insert query so that different parts are created

```
INSERT INTO table1 VALUES(100, 'b', true, 1, 1)
INSERT INTO table1 VALUES(100, 'b', true, -1, 1)
INSERT INTO table1 VALUES(100, 'b', false, 1, 2)
INSERT INTO table1 VALUES(100, 'a', true, 1, 1)
```

When i do `SELECT * from table1`

i get all the rows and value which is expected

now i write the command `OPTIMIZE table table1 final`, for collapsing to occur.

I see only these two rows

(100, 'b', false, 1, 2)
(100, 'a', true, 1, 1)

According to algorithm
![image](https://github.com/user-attachments/assets/d67aea9a-772b-45a4-9060-40dbf84c6a79)

It selects row with same primary key, so ideally shouldnt a also get collapsed ? since primary key is same ? Why didnt a collapse

**Comment by SaltTan** [comment]
Unfortunately the description of the algorithm is not correct.
'Primary key' needs to be replaced by 'ORDER BY' in this description.
In your case ClickHouse uses ORDER BY (ID, Name) to collapse rows.
